### PR TITLE
Add a "deterministic end-to-end test" runner

### LINKED ISSUE #696: Roslyn.Compilers.CompilerServer.UnitTests expects System.Collections.Immutable to have the wrong version
When 'System.Collections.Immutable' and 'System.Reflection.Metadata' don't exist in the working directory, it expects them to exist under the MSBuildToolsPath and to have the same version as the unit test assembly.

These files should either ignore the version check altogether or check against their correct version numbers.

A better solution may be to have a one time check to ensure VBCSCompiler.exe has the same version number (if it doesn't exist under the working directory) and assume the other files then have the correct version.

### LINKED ISSUE #5912: Create more thorough determinism E2E tests
We need an E2E test for the new compiler deterministic flag. The feature itself is currently broken for the E2E scenario and we need a test to validate that it is working (once the fix comes out) and continues to work moving forward.

### PR DESCRIPTION
**Pull request by TyOverby**
Closes #5912
Closes #696

### CODE REVIEW

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                {
+                    AssertEx.Fail($"Bytes were different at position { i } ({ first[i] } vs { second[i] }).  Flags used were '{ flags } { additionalFlags }'");
+                }
+            }
+        }
+
+        [Fact]
+        public void HelloWorld()
+        {
+            var source = "hello world";
```
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Get rid of this.

Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Done

**Comment by jmarolf** [comment]
test prtest/lin/dbg/unit32 please
test prtest/mac/dbg/unit32 please

**Comment by TyOverby** [comment]
@tannergooding: What was the issue number for the other bug that this one closes?

**Comment by gafter** [comment]
LGTM. I don't think this is all that is needed for #5912, though.

**Comment by TyOverby** [comment]
Review please! @dotnet/roslyn-compiler 

To make things easier on yourself, review the commits in the PR separately.  The first commit pulls some functionality out of another test that we both need and makes it available to both test runners.

The second commit actually goes through and adds the new end-to-end unit tests.

These tests take some source code, compile it using the recently-built compiler multiple times, and then make sure that the byte-for-byte output is the same.

**Comment by TyOverby** [comment]
I haven't gotten to adding the VB compiler yet.  That could be a part of this commit, or a separate one; I just wanted to get feedback on this before I fleshed everything out.

**Comment by jaredpar** [comment]
We should be adding this to the stabilization branch.  Right now this PR is targeting master.

File: src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs [hunk 2]
```diff
@@ -113,46 +49,11 @@ End Sub
...
-        };
-
-        private readonly TempDirectory _tempDirectory;
-        private readonly string _compilerDirectory;
-
-        private readonly string _csharpCompilerClientExecutable;
-        private readonly string _basicCompilerClientExecutable;
-        private readonly string _compilerServerExecutable;
-        private readonly string _buildTaskDll;
+        ExecutingTestUtils.TempExecutionEnvironment executionEnvironment = null;
```
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs]: _executionEnvironment

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
+﻿using Roslyn.Test.Utilities;
```
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Add a copyright to the file

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+using System.IO;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CompilerServer.UnitTests
+{
+    public class DeterminismTests: TestBase
+    {
+        private string flags = "/shared /deterministic+ /nologo /t:library";
```
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: _flags

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            // Setup
+            var tempDir = executionEnvironment.tempDirectory;
+            var srcFile = tempDir.CreateFile("test.cs").WriteAllText(source).Path;
+            var outFile = srcFile.Replace("test.cs", "test.dll");
+            var errorsFile = srcFile + ".errors";
+            File.Delete(executionEnvironment.compilerServerExecutable);
+
+            // Compile
+            var result = ProcessUtilities.Run("cmd", $"/C {executionEnvironment.csharpCompilerClientExecutable} { flags } { additionalFlags } { srcFile } /out:{ outFile } > { errorsFile }");
+            if (result.ExitCode != 0) {
```
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Braces on separate lines.

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                var errors = File.ReadAllText(errorsFile);
+                AssertEx.Fail($"Deterministic compile failed \n stderr: { result.Errors } \n stdout:  { errors }");
+            }
+            var bytes = File.ReadAllBytes(outFile);
+            AssertEx.NotNull(bytes);
+
+            // Cleanup
+            File.Delete(srcFile);
```
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Should we put these in a finally block?

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs (hunk 4 above)
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: We should be explicit about /pdb:none as well

**Comment by TyOverby** [comment]
@jaredpar: I think I need to delete this pull request and make a new one in order to change the target branch.  I don't want to manage two at the same time, so I'll keep this one open until I'm ready to merge.

File: src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs (hunk 2 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs]: Done.

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs (hunk 3 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Done.

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs (hunk 4 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Done, and Done.

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs (hunk 5 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Done.

File: src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs (hunk 6 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/DeterminismTests.cs]: Yeah probably.  Done.

File: src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs [hunk 7]
```diff
@@ -113,46 +49,11 @@ End Sub
...
-            {
-                var filename = Path.GetFileName(path);
-                File.Copy(path, Path.Combine(_compilerDirectory, filename));
-            }
-
-            _csharpCompilerClientExecutable = Path.Combine(_compilerDirectory, CSharpClientExeName);
-            _basicCompilerClientExecutable = Path.Combine(_compilerDirectory, BasicClientExeName);
-            _compilerServerExecutable = Path.Combine(_compilerDirectory, CompilerServerExeName);
-            _buildTaskDll = Path.Combine(_compilerDirectory, BuildTaskDllName);
+            executionEnvironment = new ExecutingTestUtils.TempExecutionEnvironment();
```
Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs]: Looks like you forgot to rename the uses.

Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs]: Yep... it looks like I caught that at exactly the same time you did.  Really should have tested before pushing.

"but it's such a simple change" he said, not seeing the red squiggles.

Comment (jaredpar) [inline src/Compilers/Core/VBCSCompilerTests/CompilerServerTests.cs]: The simple changes are always the ones that bite you.

**Comment by TyOverby** [comment]
Closed in favor of https://github.com/dotnet/roslyn/pull/6146

**Pull request closed without merging**